garden_ai/app/garden.py

```diff
@@ -168,9 +174,119 @@ def create(
     )
 
-    client.put_local_garden(garden)
+    local_data.put_local_garden(garden)
 
     if verbose:
-        metadata = client.get_local_garden(garden.uuid)
+        metadata = local_data.get_local_garden(garden.uuid)
         rich.print_json(metadata)
     return
+
+
+# TODO: allow selecting by DOI.
+@garden_app.command(no_args_is_help=True)
+def add_pipeline(
+    garden_uuid: str = typer.Option(
+        ...,
+        "-g",
+        "--garden",
+        prompt="Please enter a the name or DOI of a garden",
+        help="The name of the garden you want to add a pipeline to",
+        rich_help_panel="Required",
+    ),
+    pipeline_uuid: str = typer.Option(
+        ...,
+        "-p",
+        "--pipeline",
+        prompt="Please enter a the title or DOI of a pipeline",
+        help="The name of the pipeline you want to add",
+        rich_help_panel="Required",
+    ),
+):
+    garden_metadata = get_garden_meta(garden_uuid)
+    pipeline_uuids_already_present = set(
+        pipeline["uuid"] for pipeline in garden_metadata["pipelines"]
+    )
+    if pipeline_uuid in pipeline_uuids_already_present:
+        logger.info(f"Pipeline {pipeline_uuid} is already in Garden {garden_uuid}")
+        return
+
+    pipeline_meta = get_pipeline_meta(pipeline_uuid)
+    garden_metadata["pipelines"].append(
+        {"uuid": pipeline_uuid, "doi": pipeline_meta["doi"]}
+    )
+    local_data.put_local_garden_metadata(garden_metadata)
+    logger.info(f"Added pipeline {pipeline_uuid} to Garden {garden_uuid}")
+
+
+@garden_app.command(no_args_is_help=True)
+def publish(
+    garden_uuid: str = typer.Option(
+        ...,
+        "-g",
+        "--garden",
+        prompt="Please enter a the name or DOI of a garden",
+        help="The name of the garden you want to add a pipeline to",
+        rich_help_panel="Required",
+    ),
+):
+    # TODO: mint DOI for the Garden in this command.
+    client = GardenClient()
+
+    garden_metadata = get_garden_meta(garden_uuid)
+    pipeline_metas = [
+        get_pipeline_meta(p["uuid"]) for p in garden_metadata["pipelines"]
+    ]
+    garden_metadata["pipelines"] = pipeline_metas
+    garden_metadata["doi"] = mint_doi_from_garden_meta(garden_metadata, client)
+
+    try:
+        client.publish_garden_metadata(garden_metadata)
+    except SearchAPIError as e:
+        logger.fatal(f"Could not publish garden {garden_uuid}")
+        logger.fatal(e.error_data)
+        raise typer.Exit(code=1) from e
+
+
+# Right now we can make a Garden model from Garde JSON iff it has no pipelines.
+# Make a pipeline-less copy of the JSON so that we can make a Garden model
+# and use its DOI generating method.
+# TODO: clean up this hack once we fix Pydantic model deserialization.
+def mint_doi_from_garden_meta(garden_meta: Dict, client: GardenClient):
+    garden_copy = deepcopy(garden_meta)
+    garden_copy["pipelines"] = []
+    # No DOI currently serializes to None, which is not accepted by the Garden model.
```

Comment:
if the 'doi' key isn't in the `garden_copy` dict, there should be no errors (but this is p surprising behavior, we can/should revisit how "required" a doi should be for v0)